src/hotspot/share/gc/g1/g1ConcurrentMark.cpp

```diff
@@ -1440,6 +1440,12 @@ void G1ConcurrentMark::remark() {
       _needs_remembered_set_rebuild = (cl.total_selected_for_rebuild() > 0);
     }
 
+    if (_needs_remembered_set_rebuild) {
+      // Prune rebuild candidates based on G1HeapWastePercent.
+      // Improves rebuild time in addition to remembered set memory usage.
+      G1CollectionSetChooser::build(_g1h->workers(), _g1h->num_regions(), _g1h->policy()->candidates());
```

Comment:
The `build` method sorts according to gc efficiency and prunes.

GC efficiency can not be properly determined at this point and is entirely based on occupancy (remsets are empty); so basically what this does is sort and prune by live objects.

The whole gc-efficiency calculation can then be removed.

Actually I am not sure making the efficiency only based on free space is problematic, however we shouldn't pretend we care then.

An alternative I tried long time ago has been counting incoming references during marking as a substitute for the remset size metric. However in reality, the remembered set is a metric based on cards, so it will potentially be hard to assign cost for remembered set processing based on incoming references.